engine/flink/kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/source/flink/FlinkKafkaSource.scala

```diff
@@ -171,6 +175,45 @@ class FlinkKafkaConsumerHandlingExceptions[T](
     super.close()
   }
 
+  /**
+   * We observed that [[FlinkKafkaConsumerBase]], in `initializeState()`, may set a non-null but empty `restoredState`
+   * even though the saved state clearly contained proper state data. This makes the `open()` method treat
+   * all partitions as new ones instead of trying to fall back to offsets stored in associated consumer group.
+   *
+   * This may happen when we are changing Kafka source to a different implementation (but still a compatible one),
+   * but also when there are no changes in the scenario. There's possibly some kind of strange state incompatibility,
+   * and Flink accepts the old state from savepoint as compatible, but it restores it as empty state.
+   *
+   * It's impossible to fix Flink's source because it's deprecated and doesn't accept any changes, including bugfixes.
+   *
+   * To work around this issue we patch `restoredState` value to prevent invalid state restores and treat
+   * this situation as a new deployment without state.
+   *
+   * Note that our change may break a source reading from topics indicated by a pattern which, at the time
+   * of snapshot creation, indicates zero partitions. Nussknacker always uses concrete topic names, so for us this
+   * is acceptable.
+   */
+  private def patchRestoredState(): Unit = {
+    assert(
+      this.isInstanceOf[FlinkKafkaConsumerBase[_]],
+      s"$this must be an instance of ${classOf[FlinkKafkaConsumerBase[_]]}"
+    )
+    val restoredStateField = classOf[FlinkKafkaConsumerBase[_]].getDeclaredField("restoredState")
+    restoredStateField.setAccessible(true)
+    restoredStateField.get(this) match {
+      case null => // there is no restored stare
+      case tm: util.TreeMap[_, _] =>
+        if (tm.isEmpty) {
+          logger.info("Got empty restoredState, patching it to prevent automatic reset to the earliest offsets")
+          restoredStateField.set(this, null)
```

Comment:
How does setting this field to null prevent flink from automatic reset the consumer?

Reply:
When this field is not `null` then Flink treats all partitions that are not in state as new ones (hardcoded behavior) - it enters this branch in `open`: https://github.com/apache/flink-connector-kafka/blob/v3.1.0/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java#L578

Reply:
And when this field is set to `null` it uses specified by us resetting strategy which can be latest - here is the point of this change? I know that it might be obvious but IMO it is worth writing this.

Reply:
Added